### ISSUE TITLE
Allow the user to set the initial state of a population

### ISSUE THREAD

**Issue by celiasmith** [nengo/nengo#316]
It would be handy to be able to give a starting point (other than zero) by specifying the first population state vector.  It might even make sense to allow specifying the starting somatic voltages of the population somehow.

**Comment by Seanny123** [comment]
Any ideas on how to implement this? I'm not really familiar with the simulator, but if someone pointed me in the right direction, I'd love to take a crack at it.

**Comment by studywolf** [comment]
I think we've discussed this before, I thought we had it online but it might have been just in the office. If I remember correctly the problem here is setting up the presynaptic activity and outgoing connections of the population properly...is that right @drasmuss?

**Comment by drasmuss** [comment]
It'd be relatively easy to figure out what the optimal activities are for a population, for a given target output (and given decoders).  However, if you just set the output activites to that value, it's immediately going to diverge from that, because the population isn't getting the inputs to support that activity.  So if you really wanted that to work robustly, you'd have to back it up to all the inputs to the population (and all the inputs to those populations, etc.).  And you would probably have to fiddle with the connection signals as well, so you don't get a bunch of transients induced by the synaptic filtering.

As a first step, you could just do the first part, and put the burden on the user to initialize everything properly.  That is, if the user is wanting to set the population output to 0.5, they probably also want to initialize the inputs to that population to support that.  That wouldn't solve the synaptic filtering transients, but maybe those won't be that bad?  Who knows!

**Comment by hunse** [comment]
I think this is most useful in the case of things with recurrent connections, e.g. an integrator or an oscillator, since purely feedforward ensembles are quick to adjust their activities (i.e. little memory). For recurrent types of objects, the memory is in the synapses, so really the key would be initializing the synapses. So you're going to have to identify the recurrent connections on an ensemble that you want to initialize, which in general is tricky.

One way to constrain the problem would be to allow the user to set an initial value for a synapse. So if they wanted an integrator to have a specific initial value, they'd have to figure out what value the recurrent synapse would have, and assign that as the initial value. (This should be easy if you don't have a function on the recurrent connection, since then the target value for the recurrent population will be the same as the synapse value. If there is a function on the connection, then you have to apply that function to the synapse initial value, since the decoding happens first. We could make a helper function to do this, or even make it automatic.)

Then, if we want, we could add `initial_value` parameters to networks like Integrator and Oscillator, which will assign the initial value to the proper recurrent connection so the user doesn't have to worry about it.

**Comment by jgosmann** [comment]
This feature would be pretty useful for me to set the initial state of a memory population. I assume there has no progress been made since the last comment? Also pinging @Seanny123 who is assigned to this issue ...

**Comment by tcstewar** [comment]
No progress that I know of.  The times where I've wanted to do this, I tend to make a Node that receives input from and sends output to the Ensemble/Module/whatever whose value I want to set, and then I have that Node output the difference between what it should be and what it is (times a scaling factor).  So pretty much a classic P controller.  But if you put the scaling factor too large then you can get some wildly oscillatory behaviour.....

**Comment by tbekolay** [comment]
Another workaround for the time being is to directly set the signal values once you build the model, but before you run it. Hacky of course, but it's a workaround! The hard thing for this issue is coming up with a good API for it, IMO.

**Comment by jgosmann** [comment]
@tcstewar The problem is that this is embedded into a larger model that might react to this initialization phase in undesired ways or where there are multiple memory populations influencing each other.

@tbekolay At the moment it is not obvious to me what signals need to get initialized in what way (but I haven't thought about it much). As for the API an `initial_value` argument could be added to `nengo.Ensemble`.

**Comment by hunse** [comment]
If the thing you want to initialize is encapsulated in a network, you could run that network in a separate simulator to get the initial state, and then copy those signal values into your main simulator. This would actually be fairly easy to do programatically since signals are indexed in the builder by the model object they correspond to. You just need to make sure everything is seeded the same between both. (This is how I had always imagined the whole initialization thing happening.)

@jgosmann: The problem with `initial_value` on `Ensemble` is that it doesn't initialize synapses, as discussed above.

**Comment by jgosmann** [comment]
> If the thing you want to initialize is encapsulated in a network, you could run that network in a separate simulator to get the initial state, and then copy those signal values into your main simulator.

Would this work with nengo_ocl?

While this might be a possible workaround, it seems somewhat complicated to implement given my current evironment (e.g. the `Simulator` object is created by ctn_benchmarks, so I might need to modify that).

> The problem with initial_value on Ensemble is that it doesn't initialize synapses, as discussed above.

To me this is more of an implementation problem than what the API should look like.

Maybe the `initial_value` makes more sense on the (recurrent) connection?

**Comment by hunse** [comment]
>> The problem with initial_value on Ensemble is that it doesn't initialize synapses, as discussed above.

> To me this is more of an implementation problem than what the API should look like.

To me, they go hand in hand. If you were to set the initial value on the Ensemble, would that then initialize all connections coming out of it, too? That doesn't seem intuitive to me.

> Maybe the `initial_value` makes more sense on the (recurrent) connection?

Setting it on the recurrent connection makes more sense. It still might not fully capture things though. The membrane voltages of neurons do also contribute to the state of the system, so if you just set the state at the synapse, you might get some strange start-up transients. It might be worth investigating how significant those transients are, though. If they're not that bad, perhaps setting the state on a connection is a quick and easy feature that will address most use cases. Also, setting the initial membrane voltages of the ensemble to be random, rather than all zero, might greatly reduce these transients. 

I still think that being able to simulate sub-networks individually is the most well-defined way of doing things. That said, it might be clunky to use in practice since we would need some way to specify all the inputs to the network (perhaps we could make some way to specify node functions on the fly).

So I think ways to specify the initial membrane voltages for ensembles and the initial value of a connection (in the decoded/post-transform space) should get us really far. We might also want to consider random initial membrane voltages by default.

**Comment by jgosmann** [comment]
To me it seems that setting the state fo the synapse will me sufficient for my use case now.

> This would actually be fairly easy to do programatically since signals are indexed in the builder by the model object they correspond to.

Unfortunately, the synapse state is not stored as a signal which makes this more complicated. 

Maybe that state should also actually be stored as a signal because I ran into some problems with this before when trying to parallize the execution in the reference simulator (though there are other issues with do not make the parallelization worthwhile).

**Comment by tbekolay** [comment]
@Seanny123 and I were talking a bit about how this might be implemented, and this is what we came up with:

We could start off with a function like `set_initial_ensemble_state` in `nengo_extras` that takes in a simulator instance, an ensemble, and the value to set the initial input to... inside the function, we would find the input signal to the ensemble (something like `sim.model.sig[ens]['in']` but I'm not 100% sure about that) and set the initial value to the passed in value. Additionally, as was discussed above, we'd also have to deal with any incoming connections so that the initial value doesn't get reset to 0 by the value of an input synapse... that'll be somewhat tricky I suspect! But we should be able to get that info by looping through all of the connections in the network (`sim.model.toplevel.all_connections`) and finding any for which the `post` object is `ens` or a slice of `ens`.

**Comment by hunse** [comment]
My impression is that there's only two things in our models that really store state information over time: synapse values and neuron states. Assuming LIF neurons, we have two states to deal with, membrane voltage and refractory period. (Other neuron types are perhaps more complex.)

With regards to neuron parameters, we can do this analytically for LIF neurons by using the encoders/gains/biases to figure out the firing rates of all neurons for that input value, the inverse of which is the period between spikes. Part of that period is refractory, and part is when the voltage is changing. We can pick a random point in that firing cycle for each neuron that's firing. For neurons that aren't firing, we just set the membrane voltage to the input current. For more complex neuron types, we'd probably have to simulate the neurons for a little bit with the given input.

As for synapses, I think we should actually be worried more about outgoing connections from that population rather than incoming ones. This makes sense with something like an integrator, where you have one recurrent connection and one input connection. If you're setting the initial value of the integrator, you don't care what the input connection is doing. You're only worried about the recurrent connection, because it's an outgoing connection of the population you're setting (also an incoming one, but that's beside the point). So I think we should enumerate over all the outgoing connections from the population being set.

The simplest and most robust way to do this would again be to actually simulate the population and its outgoing connections. If we want to do it analytically for LIF neurons, and do it "right", it's pretty hard. We'd need to figure out, when we're setting the state parameters on the neurons, when each neuron last spiked, and use those spikes to figure out the state of the synapse. For a Lowpass synapse this wouldn't be too bad, but for Alpha synapses and other synapses with more than one state, I think you're essentially going to have to do a mini-simulation to get it right. If we don't care about being quite so exact (which I think is the right approach if we want to go the analytic route), we can use the activities we computed when figuring out the neuron states, pass those through the decoders/weights the outgoing connection, and set the synapse to that. 

> Unfortunately, the synapse state is not stored as a signal which makes this more complicated.

I think doing this in Nengo is probably the first step. Otherwise, whatever we build would be specific to the reference simulator, and wouldn't even work with Nengo OCL, etc.

**Comment by jgosmann** [comment]
>> Unfortunately, the synapse state is not stored as a signal which makes this more complicated.
> 
> I think doing this in Nengo is probably the first step. Otherwise, whatever we build would be specific to the reference simulator, and wouldn't even work with Nengo OCL, etc.

I agree with your statement, but I don't understand how it relates to my quoted comment.

To elaborate on my comment: What I meant to say is that it is hard to get access to the synapse's state variable in the current implementation because it is an instance variable of an object instance in a closure. I don't think there is any way to access this object instance without some tricky Python introspection. Thus a first step might be to extract this into a signal? Or at least provide access in some way, though to me it seems beneficial to have all state stored in signals. (Though then again, we probably can't ensure this in the general case involving user processes.)

I might be wrong on this (please correct me if I am), but to me it seems that setting the synapse state is much more important than the neuron state. Because the neuron state alone which is essentially the membrane voltage won't drive any firing without external input and will quickly decay to resting state. Setting the synapses, however, to a non-zero initial state will feed current input into the neurons to have them spike (even if their initial membrane voltage is at resting state, at least if the synaptic time constant is sufficient).

For me this would be most useful on recurrent memory population (which do not do pure feed-forward processing and need some time to load with the initial state that requires me to specially handle the initialization phase within my model). So from this point of view it might make sense to focus on outgoing connections, but I haven't spend much thought on focusing on incoming vs outgoing connections.

**Comment by hunse** [comment]
Yes, I meant that making the synapse state a signal is the first step.

On Jan 6, 2017 4:12 PM, "Jan Gosmann" <notifications@github.com> wrote:

Unfortunately, the synapse state is not stored as a signal which makes this
more complicated.

I think doing this in Nengo is probably the first step. Otherwise, whatever
we build would be specific to the reference simulator, and wouldn't even
work with Nengo OCL, etc.

I agree with your statement, but I don't understand how it relates to my
quoted comment.

To elaborate on my comment: What I meant to say is that it is hard to get
access to the synapse's state variable in the current implementation
because it is an instance variable of an object instance in a closure. I
don't think there is any way to access this object instance without some
tricky Python introspection. Thus a first step might be to extract this
into a signal? Or at least provide access in some way, though to me it
seems beneficial to have all state stored in signals. (Though then again,
we probably can't ensure this in the general case involving user processes.)

I might be wrong on this (please correct me if I am), but to me it seems
that setting the synapse state is much more important than the neuron
state. Because the neuron state alone which is essentially the membrane
voltage won't drive any firing without external input and will quickly
decay to resting state. Setting the synapses, however, to a non-zero
initial state will feed current input into the neurons to have them spike
(even if their initial membrane voltage is at resting state, at least if
the synaptic time constant is sufficient).

For me this would be most useful on recurrent memory population (which do
not do pure feed-forward processing and need some time to load with the
initial state that requires me to specially handle the initialization phase
within my model). So from this point of view it might make sense to focus
on outgoing connections, but I haven't spend much thought on focusing on
incoming vs outgoing connections.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/nengo/nengo/issues/316#issuecomment-271008050>, or mute
the thread
